src/Shared/Microsoft.AspNetCore.Razor.Utilities.Shared/EnumerableExtensions.cs

```diff
@@ -11,31 +12,141 @@ namespace System.Collections.Generic;
 
 internal static class EnumerableExtensions
 {
+    /// <summary>
+    ///  Projects each element of an <see cref="IEnumerable{T}"/> into a new form.
+    /// </summary>
+    /// <typeparam name="T">The type of the elements in <paramref name="source"/>.</typeparam>
+    /// <typeparam name="TResult">The type of the value returned by <paramref name="selector"/>.</typeparam>
+    /// <param name="source">A sequence of values to invoke a transform function on.</param>
+    /// <param name="selector">A transform function to apply to each element.</param>
+    /// <returns>
+    ///  Returns a new <see cref="ImmutableArray{T}"/> whose elements are the result of invoking the transform function
+    ///  on each element of <paramref name="source"/>.
+    /// </returns>
     public static ImmutableArray<TResult> SelectAsArray<T, TResult>(this IEnumerable<T> source, Func<T, TResult> selector)
     {
+        // If the source is an ImmutableArray<T> boxed as an IEnumerable<T>, it's better to unbox it here and
+        // call the SelectAsArray<T> extension method that takes an ImmutableArray<T>. Otherwise, it'll go
+        // through the IReadOnlyList<T> path.
+        if (source is ImmutableArray<T> array)
+        {
+            return ImmutableArrayExtensions.SelectAsArray(array, selector);
+        }
+
+        // If the source is an IReadOnlyList<T>, we should call the SelectAsArray<T> extension method that
+        // takes an IReadOnlyList<T>. This ensures that we don't foreach over it and incur the cost of allocating
+        // or boxing an enumerator.
+        if (source is IReadOnlyList<T> list)
+        {
+            return ReadOnlyListExtensions.SelectAsArray(list, selector);
+        }
+
+        // PERF: If we can get the count of the sequence, we can allocate the array up front.
+        if (source.TryGetCount(out var count))
+        {
+            if (count == 0)
+            {
+                return [];
+            }
+
+            var result = new TResult[count];
+
+            var index = 0;
+            foreach (var item in source)
+            {
+                result[index++] = selector(item);
+            }
+
+            Debug.Assert(result.Length == count);
```

Comment:
>Debug.Assert(result.Length == count); [](http://example.com/codeflow?start=12&length=37)

nit: Consider adding a Debug.Assert inside the loop that index isn't too large

Reply:
An `IndexOutOfRangeException` would already be thrown in that case, since `result` is created with a length of `count`.